_posts/2021-03-00-image-segmentation.md

```diff
@@ -0,0 +1,345 @@
+
+
+```python
+from dask_image import ndfilters
+
+smoothed = ndfilters.gaussian_filter(images, sigma=[0, 1, 1])
+
+```
+
+### Step 3: Segmenting objects
+
+#### Absolute threshold
+Pixels below the threshold value are background.
+
+
+```python
+absolute_threshold = smoothed > 160
+```
+
+Let's have a look at these images with napari. First we'll need to use the `%gui qt` magic:
+
+```python
+%gui qt
+```
+
+And now we can look a the images with napari:
+
+```python
+import napari
+
+viewer = napari.Viewer()
+viewer.add_image(absolute_threshold)
+viewer.add_image(images, contrast_limits=[0, 2000])
+```
+
+![Absolute threshold napari screenshot](../images/2021-image-segmentation/napari-absolute-threshold.png)
+
+#### Local threshold
+
+A better segmentation using local thresholding.
+
+
+
+```python
+thresh = ndfilters.threshold_local(smoothed, images.chunksize)
+threshold_images = smoothed > thresh
+```
+
+
+```python
+# Let's take a look at the images
+viewer.add_image(threshold_images)
+```
+
+![Local threshold napari screenshot](../images/2021-image-segmentation/napari-local-threshold.png)
+
+
+
+### Step 4: Morphological operations
+
+Morphological operations are changes we make to the shape of structures a binary image.
+
+
+https://docs.opencv.org/3.0-beta/doc/py_tutorials/py_imgproc/py_morphological_ops/py_morphological_ops.html
+
+
+
+**Erosion**
+
+![Erosion of a binary image](../images/2021-image-segmentation/erosion.png)
+
+**Dilation**
+
+![Dilation of a binary image](../images/2021-image-segmentation/dilation.png)
+
+We can combine morphological operations in different ways to get useful effects.
+
+A **morphological opening** operation is an erosion, followed by a dilation.
+
+![Morphological opening of a binary image](../images/2021-image-segmentation/opening.png)
+
+```python
+from dask_image import ndmorph
+import numpy as np
+
+structuring_element = np.array([
+    [[0, 0, 0], [0, 0, 0], [0, 0, 0]],
+    [[0, 1, 0], [1, 1, 1], [0, 1, 0]],
+    [[0, 0, 0], [0, 0, 0], [0, 0, 0]]])
+binary_images = ndmorph.binary_opening(threshold_images, structure=structuring_element)
+
+```
+
+### Step 5: Measuring objects
+Each image has many individual nuclei, so for the sake of time we'll measure a small subset of the data.
+
+
+```python
+from dask_image import ndmeasure
+
+# Create labelled mask
+label_images, num_features = ndmeasure.label(binary_images[:3], structuring_element)
+index = np.arange(num_features - 1) + 1  # [1, 2, 3, ...num_features]
+print("Number of nuclei:", num_features.compute())
+
+```
+
+    Number of nuclei: 271
+
+
+
+```python
+# Let's look at the labels
+viewer.add_labels(label_images)
+viewer.dims.set_point(0, 0)
+```
+
+Measuring objects (continued)
+
+
+```python
+# Measure objects in images
+area = ndmeasure.area(images[:3], label_images, index)
+mean_intensity = ndmeasure.mean(images[:3], label_images, index)
+```
+
+![Label image napari screenshot](../images/2021-image-segmentation/napari-label-image.png)
+
```

Comment:
"Eagle eyed readers will notice that not every nucleus is separated perfectly from its neighbours" (& add link to Nick or Juan's napari interactive segmentation tutorial?)